File: src/zig/throttle.zig
// Throttling for frame rate limiting
const std = @import("std");

pub const Throttle = struct {
    last_frame_time: i64,
    min_frame_interval: i64, // nanoseconds
    fps: u32,

    pub fn init(fps: u32) Throttle {
        const interval_ns = @divTrunc(1_000_000_000, @as(i64, fps));
        return .{
            .last_frame_time = 0,
            .min_frame_interval = interval_ns,
            .fps = fps,
        };
    }

    pub fn set_fps(self: *Throttle, fps: u32) void {
        self.fps = fps;
        self.min_frame_interval = @divTrunc(1_000_000_000, @as(i64, fps));
    }

    pub fn should_render(self: *Throttle) bool {
        const now = @as(i64, @intCast(std.time.nanoTimestamp()));
        if (now - self.last_frame_time >= self.min_frame_interval) {
            self.last_frame_time = now;
            return true;
        }
        return false;
    }

    pub fn time_until_next_frame(self: *Throttle) i64 {
        const now = @as(i64, @intCast(std.time.nanoTimestamp()));
        const elapsed = now - self.last_frame_time;
        const remaining = self.min_frame_interval - elapsed;
        return if (remaining > 0) @as(i64, @intCast(remaining)) else 0;
    }
};

File: src/native/throttle.ts
// Throttling for frame rate limiting - TypeScript implementation
// Optimized for Node.js high-resolution timing

/**
 * Throttle class for limiting render frequency.
 * Uses high-resolution time (hrtime) for accurate throttling.
 */
export class Throttle {
  private lastFrameTime: bigint = 0n;
  private minFrameInterval: bigint; // nanoseconds
  private fps: number;

  constructor(fps: number = 60) {
    this.fps = fps;
    // Calculate interval in nanoseconds: 1 second / fps
    this.minFrameInterval = BigInt(Math.floor(1_000_000_000 / fps));
  }

  /**
   * Update the target FPS
   */
  setFps(fps: number): void {
    this.fps = fps;
    this.minFrameInterval = BigInt(Math.floor(1_000_000_000 / fps));
  }

  /**
   * Check if we should render now based on throttle settings.
   * Returns true if enough time has passed since last render.
   */
  shouldRender(): boolean {
    const now = process.hrtime.bigint();
    const elapsed = now - this.lastFrameTime;

    if (elapsed >= this.minFrameInterval) {
      this.lastFrameTime = now;
      return true;
    }

    return false;
  }

  /**
   * Get time remaining until next frame can be rendered (in milliseconds).
   * Useful for scheduling the next render.
   */
  timeUntilNextFrame(): number {
    const now = process.hrtime.bigint();
    const elapsed = now - this.lastFrameTime;
    const remaining = this.minFrameInterval - elapsed;

    if (remaining <= 0) {
      return 0;
    }

    // Convert nanoseconds to milliseconds
    return Number(remaining) / 1_000_000;
  }

  /**
   * Reset the throttle (useful for testing or after long pauses)
   */
  reset(): void {
    this.lastFrameTime = 0n;
  }
}

